spec/Section 3 -- Type System.md

```diff
@@ -263,74 +263,108 @@ TypeDefinition :
+*concrete Object type*, which is always an *Object type*.
+
+An *Interface type* defines a list of fields. An type which implements this
+Interface must also implement those fields. Whenever a field claims it will
+return an *Interface type*, it will return an implementing
+*concrete Object type* during field execution.
+
+A *Union type* defines a list of possible types. Similar to Interfaces,
+whenever a field claims it will return a *Union type*, it will return one of the
+possible types as the *concrete Object type* during field execution.
+
+
+### Wrapped Types
 
 All of the types so far are assumed to be both nullable and singular: e.g. a
-scalar string returns either null or a singular string.
+scalar string returns either {null} or a singular string.
+
+:: A *wrapped type* wraps another type (its *inner type*) to create a collection
+or variation of that type. There are two kinds of wrapped types: *List type* and
+*Non-Null type*.
 
-A GraphQL schema may describe that a field represents a list of another type;
-the `List` type is provided for this reason, and wraps another type.
+:: The *inner type* is the type immediately wrapped by a *wrapped type*, which
+itself may be a *wrapped type* allowing for combinations of *wrapped type*. If
+continuously unwrapped, ultimately a *named type* is found.
 
-Similarly, the `Non-Null` type wraps another type, and denotes that the
-resulting value will never be {null} (and that a field error cannot result in a
-{null} value).
+A GraphQL schema may describe that a field or argument expects a list of another
+type; the *List type* is provided for this purpose, and wraps an *inner type*
+describing the items in the list.
 
-These two types are referred to as "wrapping types"; non-wrapping types are
-referred to as "named types". A wrapping type has an underlying named type,
-found by continually unwrapping the type until a named type is found.
+Similarly, a field or argument may describe that it must never be {null}. The
+*Non-Null type* is provided for this purpose, and wraps an *inner type*
+describing the expected type, if not {null}.
 
 
 ### Input and Output Types
 
 Types are used throughout GraphQL to describe both the values accepted as input
 to arguments and variables as well as the values output by fields. These two
-uses categorize types as *input types* and *output types*. Some kinds of types,
-like Scalar and Enum types, can be used as both input types and output types;
-other kinds of types can only be used in one or the other. Input Object types can
-only be used as input types. Object, Interface, and Union types can only be used
-as output types. Lists and Non-Null types may be used as input types or output
-types depending on how the wrapped type may be used.
+uses categorize types as an *input type* or an *output type*. Some kinds of
+types, like Scalar and Enum types, can be used as both input types and output
+types; other kinds of types can only be used in one or the other.
```

Comment:
*A little off-topic* Whilst we're on this subject... generally `scalar` types are suitable for input and output, but in some cases (for example the `Upload` scalar used by the GraphQL Upload Spec) a custom scalar may only be suitable for one but not the other. Is this specifically disallowed, or do we need to be more cautious with our wording?

Reply:
We actually don't have the tools to assert this restriction right now. We don't specifically disallow it, but we also can't enforce it (with validation - a custom scalar could just fail coercion, but thats a bit of a hack IMO)

Reply:
Yep; I’m happy keeping the status quo; just felt it was worth mentioning.

Reply:
> We actually don't have the tools to assert this restriction right now

I'd be interested in digging into this distinction for a later release: I have some usages around input vs output enums that I'd like to add clarity to. As the spec stands, [this statement](https://spec.graphql.org/draft/#sel-GAHfRDABABvEomB):
> GraphQL services must return one of the defined set of possible values. If a reasonable coercion is not possible they must raise a field error.

can mean it's actually a breaking change to add a new enum value to an existing enum  type (we make no statement about avoiding exhaustive switches on output enums, for example). We don't get into details in the spec itself on what a breaking type system change is, with just a [short blurb about type system evolution](https://spec.graphql.org/draft/#sec-Validation.Type-system-evolution).

Reply:
https://github.com/graphql/graphql-js/issues/968#issuecomment-317834310

Reply:
Relevant